### DISCUSSION TITLE
Can I create a port that has different interface and implementation dependencies

### DISCUSSION THREAD

**Discussion by tonygould** [microsoft/vcpkg#31015] [General]
If the project P I want to publish only uses another project I for its implementation, how do I stop a consumer C of project P bringing in and building I? This is important for binary caching on CI, where in general the binary cache for P will already be available, so it would not in theory be necessary to bring in I.

As a case in point, a small part of our large project uses boost, header-only, and as boost is only used for implementation, boost header files aren't used on our project interface headers. But due to the monolithic nature of boost, any client project has to bring in ~60 boost ports even though it doesn't need them. This is time consuming and noisy.

In our registry, we have a vcpkg.json in our port directory (alongside our portfile.cmake) for our project P, but my understanding is that boost needs to be listed here, because it's required for _building_ our project. But how do I stop a consumer of project P bringing in the 60-odd boost ports when it's been built previously and a binary cache is available?

We're using our own registry, but the question would also apply to the public vcpkg registry.

**Comment by autoantwort** [comment]
This is currently not possible. vcpkg does not know which dependency is only used at build time and which at build and runtime.

**Reply by autoantwort** [reply]
I guess since it only makes sense for header only libs it was never implemented.

**Reply by tonygould** [reply]
Thank you for your reply.

It would also make sense for building a shared object or a dll, where the dependency is a static library and is build into the resultant executable.

With a large project split into sub-projects, the ability to create an abstraction layer becomes quite important -- just like having private and public in a class. We haven't hit the issue until recently -- now we've been using vcpkg in a commercial environment with several teams of developers for some time, and it has started to become an issue, and it will grow in importance.

**Comment by dg0yt** [comment]
Note that the legal notes (aka `copyright`) for the header-only ports are installed only with the ports' artifacts. You don't get a full set unless you install them all.
